File: src/ipc/failure.ts
/// Reading a backend failure.
///
/// Commands used to reject with a bare string, so the interface told failures apart by matching
/// human-readable text — and rewording a sentence could change what the window did. The refactored
/// commands reject with `{ code, message }` instead: the code is a stable contract, the message is
/// the only text meant for a person.
///
/// Both shapes reach here, because the backend is being converted a workflow at a time. Anything
/// that is still a string is reported as `internal`, which is exactly what it is: a failure the
/// backend has not classified yet.
import type { ErrorCode } from "../types";

export type Failure = {
  code: ErrorCode;
  message: string;
};

/// Normalise anything a rejected `invoke` can produce into a failure with a code.
///
/// Never throws, and never returns an empty message: an interface that renders "[object Object]"
/// or an empty line has lost the failure entirely, which is worse than a vague sentence.
export function failure(error: unknown): Failure {
  if (typeof error === "string") {
    return { code: "internal", message: error };
  }
  if (
    error &&
    typeof error === "object" &&
    "message" in error &&
    typeof (error as Failure).message === "string"
  ) {
    const found = error as Failure;
    return {
      code: typeof found.code === "string" ? found.code : "internal",
      message: found.message,
    };
  }
  return {
    code: "internal",
    message: "Something went wrong. Check the log for details.",
  };
}

/// The sentence to show for a rejected `invoke`.
///
/// The common case: most call sites want the text, not the code. Use `failure` directly where the
/// interface has to behave differently for a particular code.
export function message(error: unknown): string {
  return failure(error).message;
}

File: src-tauri/src/failure.rs
//! Turning a backend failure into something the window can act on.
//!
//! The frontend used to tell failures apart by matching human-readable text, so rewording a
//! sentence could change what the interface did. A [`Failure`] carries a stable `code` alongside
//! the sentence, and React switches on the code.
//!
//! # Privacy
//!
//! Only `code` and `message` cross to the window. The internal `cause` is logged here and never
//! serialised: it is written for a developer, and the window may be photographed.

use orbiter_core::domain::errors::{ErrorCode, OperationError};

/// A backend failure, as the window receives it.
///
/// Serialises as `{ "code": "review_stale", "message": "…" }`. The code is a contract with the
/// frontend; changing one is a change to both sides.
#[derive(serde::Serialize)]
pub struct Failure {
    /// Stable, machine-readable classification.
    pub code: ErrorCode,
    /// The sentence a person reads. The only human-readable text that crosses.
    pub message: String,
}

impl From<OperationError> for Failure {
    /// Drop the internal cause into the log and keep what the window may see.
    fn from(error: OperationError) -> Self {
        if let Some(cause) = &error.cause {
            tracing::info!(code = error.code.as_str(), detail = %cause, "operation failed");
        }
        Self {
            code: error.code,
            message: error.message,
        }
    }
}

impl From<String> for Failure {
    /// Carry a message from a workflow this refactor has not reached yet.
    ///
    /// Everything still returning `Result<_, String>` arrives here and is classified as
    /// [`ErrorCode::Internal`]. Each use is a place the structured errors have yet to reach, so
    /// it is worth being able to find them.
    fn from(message: String) -> Self {
        Self {
            code: ErrorCode::Internal,
            message,
        }
    }
}

impl From<&str> for Failure {
    /// Carry a borrowed message, as above.
    fn from(message: &str) -> Self {
        Self::from(message.to_owned())
    }
}

/// A failure with no better classification.
///
/// Used by the few adapters that still bridge string-returning code.
pub fn internal(message: impl Into<String>) -> Failure {
    Failure {
        code: ErrorCode::Internal,
        message: message.into(),
    }
}
